### PR TITLE
toJSON/fromJSON for transactions

### PR DESCRIPTION
**Pull request by sunify**

### CODE REVIEW

File: lib/input.js [hunk 1]
```diff
@@ -78,7 +78,8 @@ export default class Input {
       };
     }
     if (this.isSpend()) {
-      const input = this.prevout;
+      const input = { ...this.prevout }; // toJSON shouldn't mutate existing objects
```
Comment (troggy) [inline lib/input.js]: good catch!

File: lib/input.js [hunk 2]
```diff
@@ -103,7 +104,12 @@ export default class Input {
     if (json.hash) {
       const input = new Input(Outpoint.fromJSON(json));
       if (json.signer) {
-        input.setSig(json.r, json.s, json.v, json.signer);
+        input.setSig(
+          new Buffer(json.r.replace('0x', ''), 'hex'),
```
Comment (troggy) [inline lib/input.js]: Also `toBuffer` from `ethereumjs-util` can be used for this operation. Like this: `toBuffer(json.r)`

https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/index.md#tobuffer

File: lib/input.js [hunk 3]
```diff
@@ -78,7 +78,8 @@ export default class Input {
       };
     }
     if (this.isSpend()) {
-      const input = this.prevout;
+      const input = { ...this.prevout }; // toJSON shouldn't mutate existing objects
+      input.hash = `0x${input.hash.toString('hex')}`;
```
Comment (troggy) [inline lib/input.js]: `bufferToHex` from `ethereumjs-util` can be used for this operation. Like this: `bufferToHex(input.hash)`

https://github.com/ethereumjs/ethereumjs-util/blob/master/docs/index.md#buffertohex

File: lib/input.js (hunk 2 above)
Comment (sunify) [reply lib/input.js]: Thanks!

**Pull request merged**